File: test1.py
import cv2
import numpy as np 
import time
from skimage.metrics import structural_similarity

cap = cv2.VideoCapture(0)


def nothing(x):
    pass

def srch():
    pass
    

#Controler setup
img = np.zeros((128,255), np.uint8)
cv2.namedWindow('image')
cv2.createTrackbar('x','image',0,320*2-64,nothing)
cv2.createTrackbar('y','image',0,480-64,nothing)
cv2.createTrackbar('Thresh','image',0,255,nothing)
cv2.createTrackbar('Thresh2','image',0,255,nothing)
cv2.createTrackbar('ksize','image',0,15,nothing) 

def controler():
    
    x = cv2.getTrackbarPos('x','image')
    y = cv2.getTrackbarPos('y','image')
    Thr = cv2.getTrackbarPos('Thresh','image')
    Thr2 = cv2.getTrackbarPos('Thresh2','image')
    krnlsize = cv2.getTrackbarPos('ksize','image')
    
    return x,y,Thr,Thr2,krnlsize

def cagedraw(frame):
    
    
    [x,y,_,_,_]=controler()
    line_thickness=2
    size=128
    x1a, y1a= -640,0+y
    x2a, y2a= 640, 0+y
    
    x1b, y1b= -640,0+size+y
    x2b, y2b= 640, 0+size+y
    
    x1c, y1c= x-0,-480
    x2c, y2c= x+0, 480
    
    x1d, y1d= x+size, -480
    x2d, y2d= x+size, 480
    
    framecop=frame.copy()
    cv2.line(framecop, (x1a,y1a), (x2a,y2a), (255,255,255),thickness=line_thickness) 
    cv2.line(framecop, (x1b,y1b), (x2b,y2b), (255,255,255),thickness=line_thickness) 
    cv2.line(framecop, (x1c,y1c), (x2c,y2c), (255,255,255),thickness=line_thickness)
    cv2.line(framecop, (x1d,y1d), (x2d,y2d), (255,255,255),thickness=line_thickness)
    box=framecop[y:y+size,x:x+size]

    
    return framecop,box

def show(frame):
    
    
    #cv2.imshow("Frame",frame)
    [framecop,box]=cagedraw(frame)
    cv2.imshow("Frame+cage",framecop)
    #cv2.imshow("Box",box)
    if cv2.waitKey(1) == ord('w'):
        yesyes=1
        cage=box.copy()
        grayboxpic=cv2.cvtColor(cage, cv2.COLOR_BGR2GRAY)
        cv2.imshow("grayboxpic", grayboxpic)
        
        while yesyes:
            ret, frame = cap.read()
            [framecop,box]=cagedraw(frame)
            cv2.imshow("Frame+cage",framecop)
            if cv2.waitKey(1) == ord('t'):
                glavni()
            if cv2.waitKey(1) == ord('e'):
                [x,y,Thr,_,_]=controler()
                xx=x
                yy=y
                size=128
                highscore=1
                r=8
                step=4
                ro=r
                line_thickness=2
                xxc=0
                while highscore <= 75:
                    
                    ret, frame = cap.read() #original picture
                    gray = cv2.cvtColor(frame, cv2.COLOR_BGR2GRAY) ##RGB to gray
                    newgray = gray.copy()
                    cv2.imshow('all',newgray)
                    if cv2.waitKey (1) == ord('q'):
                        cap.release()
                        cv2.destroyAllWindows()
                        glavni()
                        break
                    for yyc in range(-r,r+ro,step):
                        
                        
                        newgraycage = gray.copy()
                        
                        x1a, y1a= -640,0+yy-yyc
                        x2a, y2a= 640, 0+yy-yyc
                        
                        x1b, y1b= -640,0+size-yyc+yy
                        x2b, y2b= 640, 0+size-yyc+yy
                        
                        x1c, y1c= xxc-0+xx,-480
                        x2c, y2c= xxc+0+xx, 480
                        
                        x1d, y1d= xxc+size+xx, -480
                        x2d, y2d= xxc+size+xx, 480
                        
                        newgray = gray.copy()
                        cv2.line(newgray, (x1a,y1a), (x2a,y2a), (255,255,255),thickness=line_thickness) 
                        cv2.line(newgray, (x1b,y1b), (x2b,y2b), (255,255,255),thickness=line_thickness) 
                        cv2.line(newgray, (x1c,y1c), (x2c,y2c), (255,255,255),thickness=line_thickness)
                        cv2.line(newgray, (x1d,y1d), (x2d,y2d), (255,255,255),thickness=line_thickness)
                        
                        ret, frame = cap.read() #original picture
                        gray = cv2.cvtColor(frame, cv2.COLOR_BGR2GRAY)
                        cv2.imshow('all',newgray)
                        
                        if cv2.waitKey (1) == ord('q'):
                            cap.release()
                            cv2.destroyAllWindows()
                            glavni()
                            break
                        ret, frame = cap.read() #original picture
                        gray = cv2.cvtColor(frame, cv2.COLOR_BGR2GRAY)
                        
                        for xxc in range(-r,r+ro,step):
                            
                            
                            newgraycage = gray.copy()
                            
                            x1a, y1a= -640,0+yy-yyc
                            x2a, y2a= 640, 0+yy-yyc
                            
                            x1b, y1b= -640,0+size-yyc+yy
                            x2b, y2b= 640, 0+size-yyc+yy
                            
                            x1c, y1c= xxc-0+xx,-480
                            x2c, y2c= xxc+0+xx, 480
                            
                            x1d, y1d= xxc+size+xx, -480
                            x2d, y2d= xxc+size+xx, 480
                            
                            newgray = gray.copy()
                            cv2.line(newgray, (x1a,y1a), (x2a,y2a), (255,255,255),thickness=line_thickness) 
                            cv2.line(newgray, (x1b,y1b), (x2b,y2b), (255,255,255),thickness=line_thickness) 
                            cv2.line(newgray, (x1c,y1c), (x2c,y2c), (255,255,255),thickness=line_thickness)
                            cv2.line(newgray, (x1d,y1d), (x2d,y2d), (255,255,255),thickness=line_thickness)
                            
                            #Filter box
                            graybox=newgray[yy-yyc:yy-yyc+size,xx+xxc:xx+xxc+size]
                            
                            ret,thresh1new = cv2.threshold(graybox,Thr,255,0)
                            cv2.imshow("Grayuncage",graybox)
                            cv2.imshow('all',newgray)
                            #cv2.imshow('Grayboxpic',grayboxpic)
                            
                            
                            (score, diff) = structural_similarity(graybox, grayboxpic, full=True)
                            if score*100 > 75:
                                print("Found a match")
                                highscore=score*100
                                yesyes=0
                                cv2.imshow("Grayuncagefinal",graybox)
                            print(score*100)
                            #diff = 255 - (diff * 255).astype("uint8")q
                            #print("Image Similarity: {:.4f}%".format(score * 100))
                            
                                
                                
                    
                    r=r+step
                yesyes=0
            
        
        

def glavni():
    while True:
        ret, frame = cap.read()
        show(frame)

        if cv2.waitKey (1) == ord('q'):
            cap.release()
            cv2.destroyAllWindows()
            break
        
                
        
        



glavni()

cap.release()
cv2.destroyAllWindows()

File: test1.cpp
#include <opencv2/opencv.hpp>
#include <iostream>
#include <stdio.h>

using namespace cv;
using namespace std;

int main(int argc, char** argv)
{
VideoCapture cap(0);
if (!cap.isOpened())
{
cout << "Error opening video stream or file" << endl;
return -1;
}
namedWindow("image", WINDOW_NORMAL);
createTrackbar("x", "image", 0, 320 * 2 - 64, NULL);
createTrackbar("y", "image", 0, 480 - 64, NULL);
createTrackbar("Thresh", "image", 0, 255, NULL);
createTrackbar("Thresh2", "image", 0, 255, NULL);
createTrackbar("ksize", "image", 0, 15, NULL);

while (1)
{
    Mat frame;
    cap.read(frame);

    int x = getTrackbarPos("x", "image");
    int y = getTrackbarPos("y", "image");
    int Thr = getTrackbarPos("Thresh", "image");
    int Thr2 = getTrackbarPos("Thresh2", "image");
    int krnlsize = getTrackbarPos("ksize", "image");

    line_thickness = 2;
    size = 128;
    int x1a, y1a = -640, 0 + y;
    int x2a, y2a = 640, 0 + y;

    int x1b, y1b = -640, 0 + size + y;
    int x2b, y2b = 640, 0 + size + y;

    int x1c, y1c = x - 0, -480;
    int x2c, y2c = x + 0, 480;

    int x1d, y1d = x + size, -480;
    int x2d, y2d = x + size, 480;

    Mat framecop = frame.clone();
    line(framecop, Point(x1a, y1a), Point(x2a, y2a), Scalar(255, 255, 255), line_thickness);
    line(framecop, Point(x1b, y1b), Point(x2b, y2b), Scalar(255, 255, 255), line_thickness);
    line(framecop, Point(x1c, y1c), Point(x2c, y2c), Scalar(255, 255, 255), line_thickness);
    line(framecop, Point(x1d, y1d), Point(x2d, y2d), Scalar(255, 255, 255), line_thickness);
    Mat box = framecop(Rect(x, y, size, size));

    imshow("image", framecop);
    imshow("box", box);

    char c = (char)waitKey(25);
    if (c == 27)
        break;
}

cap.release();
destroyAllWindows();

return 0;
}
void show(Mat frame) {
Mat framecop, box;
vector<Mat> res = cagedraw(frame);
framecop = res[0];
box = res[1];

imshow("Frame+cage", framecop);

if (waitKey(1) == 'w') {
    int yesyes = 1;
    Mat cage = box.clone();
    Mat grayboxpic;
    cvtColor(cage, grayboxpic, COLOR_BGR2GRAY);
    imshow("grayboxpic", grayboxpic);
    
    while (yesyes) {
        Mat frame;
        cap >> frame;
        
        vector<Mat> res = cagedraw(frame);
        framecop = res[0];
        box = res[1];
        
        imshow("Frame+cage", framecop);
        
        if (waitKey(1) == 't') {
            glavni();
        }
        if (waitKey(1) == 'e') {
            vector<int> control = controler();
            int x = control[0], y = control[1], Thr = control[2];
            int xx = x, yy = y;
            int size = 128;
            int highscore = 1;
            int r = 8, step = 4, ro = r, line_thickness = 2, xxc = 0;
            while (highscore <= 75) {
                Mat frame;
                cap >> frame;
                Mat gray;
                cvtColor(frame, gray, COLOR_BGR2GRAY);
                Mat newgray = gray.clone();
                imshow("all", newgray);
                if (waitKey(1) == 'q') {
                    cap.release();
                    destroyAllWindows();
                    glavni();
                    break;
                }
                for (int yyc = -r; yyc <= r + ro; yyc += step) {
                    Mat newgraycage = gray.clone();
                    int x1a = -640, y1a = 0 + yy - yyc;
                    int x2a = 640, y2a = 0 + yy - yyc;
                    int x1b = -640, y1b = 0 + size - yyc + yy;
                    int x2b = 640, y2b = 0 + size - yyc + yy;
                    int x1c = xxc - 0 + xx, y1c = -480;
                    int x2c = xxc + 0 + xx, y2c = 480;
                    int x1d = xxc + size + xx, y1d = -480;
                    int x2d = xxc + size + xx, y2d = 480;
                    newgray = gray.clone();
                    line(newgray, Point(x1a, y1a), Point(x2a, y2a), Scalar(255, 255, 255), line_thickness);
                    line(newgray, Point(x1b, y1b), Point(x2b, y2b), Scalar(255, 255, 255), line_thickness);
                    line(newgray, Point(x1c, y1c), Point(x2c, y2c), Scalar(255, 255, 255), line_thickness);
                    line(newgray, Point(x1d, y1d), Point(x2d, y2d), Scalar(255, 255, 255), line_thickness);
                    Mat frame;
                    cap >> frame;
                    cvtColor(frame, gray, COLOR_BGR2GRAY);
                    imshow("all", newgray);
                    void glavni() {
while (true) {
Mat frame;
cap.read(frame);
show(frame);
    if (waitKey(1) == 'q') {
        cap.release();
        destroyAllWindows();
        break;
    }
}
}

glavni();

cap.release();
destroyAllWindows();





                   
